<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>blast &mdash; cctk 0.5.3 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="constrain" href="constrain.html" />
    <link rel="prev" title="CRISPRtree" href="CRISPRtree.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> cctk
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="CRISPRdiff.html">CRISPRdiff</a></li>
<li class="toctree-l1"><a class="reference internal" href="CRISPRtree.html">CRISPRtree</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">blast</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#before-you-run">Before you run</a></li>
<li class="toctree-l2"><a class="reference internal" href="#basic-usage">Basic Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output-files">Output files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#crispr-spacers-fna">CRISPR_spacers.fna</a></li>
<li class="toctree-l3"><a class="reference internal" href="#array-ids-txt">Array_IDs.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="#array-seqs-txt">Array_seqs.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="#array-locations-bed">Array_locations.bed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#array-network-txt">Array_network.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="#crispr-summary-table-csv">CRISPR_summary_table.csv</a></li>
<li class="toctree-l3"><a class="reference internal" href="#crispr-summary-table-txt">CRISPR_summary_table.txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="#spacer-cluster-members-txt">Spacer_cluster_members.txt</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#advanced-usage">Advanced Usage</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#associating-arrays-from-different-contigs">Associating arrays from different contigs</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#example-regex-or-id-file">Example regex or ID file</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#limitations-and-considerations">Limitations and considerations</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="constrain.html">constrain</a></li>
<li class="toctree-l1"><a class="reference internal" href="evolve.html">evolve</a></li>
<li class="toctree-l1"><a class="reference internal" href="minced.html">minced</a></li>
<li class="toctree-l1"><a class="reference internal" href="network.html">network</a></li>
<li class="toctree-l1"><a class="reference internal" href="spacerblast.html">spacerblast</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">cctk</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>blast</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/content/blast.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="blast">
<h1>blast<a class="headerlink" href="#blast" title="Permalink to this headline"></a></h1>
<section id="introduction">
<span id="blast-intro"></span><h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> identifies CRISPR arrays in sequences using BLASTN to search for one or more repeat sequences. It processes the results of that BLAST search through the following steps to construct arrays:</p>
<ol class="arabic">
<li><p>For each hit, extend the sequence to cover the full query length if BLAST returned a shorter match.</p></li>
<li><p>Calculate percent identity between the relevant repeat and the (now full length) blast hit and discard the hit if it does not reach the set threshold (set using the <code class="docutils literal notranslate"><span class="pre">-c</span></code> option).</p></li>
<li><p>Sort the remaining BLAST hits by:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>The sequence in which they were identified,</p></li>
<li><p>The strand on which they were identified, and</p></li>
<li><p>The start location.</p></li>
</ol>
</div></blockquote>
</li>
<li><p>Work through the list of ordered BLAST hits, using the following steps to consider each hit and associate hits together into a single array:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Is this hit closer than some threshold distance from the last hit (set using <code class="docutils literal notranslate"><span class="pre">-i</span></code>; default 80)?</p></li>
<li><p>Is this hit on the same strand as the last hit?</p></li>
<li><p>Is this hit on the same contig as the last hit?</p></li>
</ol>
</div></blockquote>
</li>
<li><p>Once the set of repeats in each array is determined, the sequence in between these repeats (CRISPR spacers) is retrieved in batches using <code class="docutils literal notranslate"><span class="pre">blastdbcmd</span></code></p></li>
</ol>
<p>An important feature of this process is that <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> has no access to the filenames or other information that connects contigs together. Therefore, in order to relate different arrays found in the same assembly, <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> requires information about which information in sequence headers relates contigs together. This is described below in the <a class="reference internal" href="#blast-basic"><span class="std std-ref">Basic Usage</span></a> section.</p>
</section>
<section id="before-you-run">
<span id="blast-before-you-run"></span><h2>Before you run<a class="headerlink" href="#before-you-run" title="Permalink to this headline"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> requires you to provide the sequences you wish to search in the form of a blast database. This can be acheived simply using the <code class="docutils literal notranslate"><span class="pre">makeblastdb</span></code> from NCBI BLAST+ command which is included in the conda installation of CCTK. However, before making your blastdb, please confirm that your sequences meet the following requirements:</p>
<ol class="arabic simple">
<li><p>No pipe symbols (“|”) in any of your fasta headers.</p></li>
<li><p>None of the fasta headers in the sequences are the same.</p></li>
<li><p>If your sequences are broken up into multiple contigs, ensure that each fasta header contains an identifier that can be used to associate the sequences.</p></li>
</ol>
<p>If you need to modify your fasta headers to add an identifier then a simple option is to add the filename to every header it contains (assuming your filenames are informative). <strong>N.B.</strong> It is a good idea to make a copy of your files before modifying them in case something goes wrong!</p>
<p>This modification can be acheived with something like the following bash loop (run in the directory with your sequence files):</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> file <span class="k">in</span> *<span class="p">;</span> <span class="k">do</span> <span class="nv">id</span><span class="o">=</span><span class="si">${</span><span class="nv">file</span><span class="p">%.*</span><span class="si">}</span><span class="p">;</span> sed -i <span class="s2">&quot;s/&gt;/&gt;</span><span class="si">${</span><span class="nv">id</span><span class="si">}</span><span class="s2">_/&quot;</span> <span class="nv">$file</span><span class="p">;</span> <span class="k">done</span>
</pre></div>
</div>
<p>This will add the filename (minus extension) to the beginning of each fasta header in each file.</p>
<p>Making a blastdb from a directory containing your sequences to search can be acheived in two steps (if you are only searching a single sequence file skip the first step). Assuming your sequences are in a directory called sequences/:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>cat sequences/* &gt; all_seqs.fna
makeblastdb -in all_seqs.fna -out seq_blastdb -dbtype nucl -parse_seqids
</pre></div>
</div>
<p><strong>N.B.</strong> It is essential to include the <code class="docutils literal notranslate"><span class="pre">-parse_seqids</span></code> option when creating the blastdb. You can be check if your blastdb was made using <code class="docutils literal notranslate"><span class="pre">-parse_seqids</span></code> by looking for files ending in .nog and .nos. If those files exist associated with your blastdb then it was made correctly.</p>
</section>
<section id="basic-usage">
<span id="blast-basic"></span><h2>Basic Usage<a class="headerlink" href="#basic-usage" title="Permalink to this headline"></a></h2>
<p>The minimal command to run <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> requires 3 inputs: a BLASTdb, a fasta-format file containing one or more CRISPR repeat sequences, and a directory into which files should be written. If you only want CRISPR spacers and array membership information, this is sufficient. However, if you want information about which assemblies contain each array you will need to provide additional information (see the <a class="reference internal" href="#blast-advanced"><span class="std std-ref">Advanced Usage</span></a> section).</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>cctk blast -d &lt;BLASTdb&gt; -r &lt;repeats file&gt; -o &lt;output dir&gt;
</pre></div>
</div>
<p><strong>N.B.</strong> The orientation of the CRISPR repeats in your input file will be used to orient the CRISPR spacers output by <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code>. If spacer and array orientation are important for the downstream analysis of your data (for example if you intend to use <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">CRISPRtree</span></code> or <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">spacerblast</span></code>), please ensure that your repeats are oriented such that the leader end of the array will be 5’ of your repeat and the trailer end will be 3’.</p>
<p><strong>N.B.2</strong> <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> is able to perform both <code class="docutils literal notranslate"><span class="pre">blastn</span></code> and processing steps using multiple threads to reduce running time. You can specify the number of threads to use with <code class="docutils literal notranslate"><span class="pre">-t</span></code></p>
</section>
<section id="output-files">
<h2>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline"></a></h2>
<p>Both <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> and <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> produce the same output files. <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> writes several output files to the directory you specify. In addition to creating output files, a summary of the run is written to stderr stating the number of unique spacers that were identified and the number of distinct arrays in which they were identified.</p>
<section id="crispr-spacers-fna">
<span id="blast-crispr-spacers"></span><h3>CRISPR_spacers.fna<a class="headerlink" href="#crispr-spacers-fna" title="Permalink to this headline"></a></h3>
<p><strong>Summary</strong></p>
<p>Sequences of all of the unique spacers that were identified in the provided assemblies. Fasta headers are constructed from the best matching repeat ID and an integer that counts the number of spacers found associated with that repeat. See <a class="reference internal" href="minced.html#minced-advanced"><span class="std std-ref">Advanced Usage</span></a> for details about how the repeat ID to assign to each spacer is chosen and how orientation of the spacer sequence is chosen by <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code>. <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> gets the repeat ID portion of the spacer fasta header from the fasta header of the repeat used to identify the CRISPR array by BLAST.</p>
<p><strong>Format</strong></p>
<p>Fasta nucleotide sequence</p>
<p><strong>Example</strong></p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>&gt;1F_1
GGTACGTGGTTTCGACCAACAGCACTGCCCAA
&gt;1F_2
AGGCTGCCAAGTCGGTGCGCGAGGCCGGCTTT
&gt;1F_3
TGCAGCGATTGCACCTTGGCCTGCTGCCGATC
&gt;1E_1
CATCTGGCCGGGGCTCGGGTCTGGTTCTACGA
&gt;1E_2
GATGGCAACCGGCGTTTGTCCGCGCTGAACTG
</pre></div>
</div>
</section>
<section id="array-ids-txt">
<span id="blast-array-ids"></span><h3>Array_IDs.txt<a class="headerlink" href="#array-ids-txt" title="Permalink to this headline"></a></h3>
<p><strong>Summary</strong></p>
<p>CRISPR arrays are defined as distinct by <code class="docutils literal notranslate"><span class="pre">cctk</span></code> if they contain a single different spacer to any other arrays. Each distinct array is assigned a numerical ID based on the order in which they are found in input assemblies. This file contains the IDs of the spacers in each array.</p>
<p><strong>Format</strong></p>
<p>2 columns, tab-delimited.</p>
<p>Column 1: ID of array
Column 2: Space-delimited list of IDs (fasta headers) of spacers in this array</p>
<p><strong>Example</strong></p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="m">1</span>       1F_42 1F_18 1F_153 1F_53 1F_82 1F_148
<span class="m">2</span>       1E_90 1E_56 1E_166 1E_26 1E_141 1E_13 1E_77
</pre></div>
</div>
</section>
<section id="array-seqs-txt">
<span id="blast-array-seqs"></span><h3>Array_seqs.txt<a class="headerlink" href="#array-seqs-txt" title="Permalink to this headline"></a></h3>
<p><strong>Summary</strong></p>
<p>This file contains the sequence of the spacers in each array.</p>
<p><strong>Format</strong></p>
<p>2 columns, tab-delimited.</p>
<p>Column 1: ID of array
Column 2: Space-delimited list of sequence of spacers in this array</p>
<p><strong>Example</strong></p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="m">1</span>       GGTACGTGGTTTCGACCAACAGCACTGCCCAA AGGCTGCCAAGTCGGTGCGCGAGGCCGGCTTT
<span class="m">2</span>       CATCTGGCCGGGGCTCGGGTCTGGTTCTACGA GATGGCAACCGGCGTTTGTCCGCGCTGAACTG
</pre></div>
</div>
</section>
<section id="array-locations-bed">
<span id="blast-array-locations"></span><h3>Array_locations.bed<a class="headerlink" href="#array-locations-bed" title="Permalink to this headline"></a></h3>
<p><strong>Summary</strong></p>
<p>Contig names and contig locations in which CRISPR arrays were identified.</p>
<p><strong>Format</strong></p>
<p>BED format.</p>
<p>First line is a “#” character followed by tab-delimited column names.</p>
<p>Name column contains the ID of the array at the indicated location. This ID corresponds to the IDs in <a class="reference internal" href="minced.html#array-ids"><span class="std std-ref">Array_IDs.txt</span></a> and <a class="reference internal" href="minced.html#array-seqs"><span class="std std-ref">Array_seqs.txt</span></a></p>
<p><strong>Example</strong></p>
<p>N.B. when viewing this file in a text editor, the headings and column contents will usually not line up, visually. If you wish to view this file for manual inspection, it will read into excel with proper column assignments or can be viewed in the terminal using <code class="docutils literal notranslate"><span class="pre">column</span> <span class="pre">-t</span> <span class="pre">Array_locations.bed</span> <span class="pre">|</span> <span class="pre">less</span></code></p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1">#contig contigStart     contigEnd       name    score   strand</span>
Assembly1_contig2       <span class="m">208444</span>  <span class="m">209013</span>  <span class="m">6</span>       <span class="m">0</span>       -
Assembly1_contig6       <span class="m">19991</span>   <span class="m">20559</span>   <span class="m">7</span>       <span class="m">0</span>       +
Assembly2_contig1       <span class="m">29424</span>   <span class="m">30050</span>   <span class="m">11</span>      <span class="m">0</span>       -
</pre></div>
</div>
</section>
<section id="array-network-txt">
<span id="blast-array-network"></span><h3>Array_network.txt<a class="headerlink" href="#array-network-txt" title="Permalink to this headline"></a></h3>
<p><strong>Summary</strong></p>
<p>Network representation of the number and proportion of spacers that arrays have in common with one another. Each pair of arrays that share one or more spacers are respresented by an edge in the network. The similarity between arrays is represented as both the number of spacers in common, and the Jaccard similarity index of the two arrays. The repeat ID associated with each array is also included.</p>
<p>This file can be easily read into a network visualization software such as cytoscape, as demonstrated in the <a class="reference external" href="tutorial.html">tutorial</a>.</p>
<p>Jaccard similarity between two arrays is defined as the number of unique spacers in common between the two arrays, divided by the combined number of unique spacers present in the two arrays.</p>
<p>e.g. for the following 2 arrays (as they would be represented in Array_IDs.txt):</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Array   Spacers
<span class="m">1</span>       1F_1 1F_2 1F_3
<span class="m">2</span>       1F_4 1F_2 1F_3
</pre></div>
</div>
<p>The array both contain spacers 1F_2 and 1F_3, while each array also contains one spacer that is not present in the other array. Therefore, the 2 shared spacers are 1F_2 and 1F_3, while the list of 4 total unique spacers in the two arrays is 1F_1, 1F_2, 1F_3, and 1F_4. This results in a Jaccard similarity index of 2/4 = 0.5</p>
<p>Jaccard is an effective similarity measure for comparing CRISPR arrays as it takes into account both the number of spacers in common between two arrays, and the spacers present in each array that are not shared.</p>
<p><strong>Format</strong></p>
<p>Tab-delimited.</p>
<p>First line is header information</p>
<p><strong>Example</strong></p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Array_A Array_B Shared_spacers  Jaccard_similarity      Array_A_type    Array_B_type
<span class="m">6</span>       <span class="m">4</span>       <span class="m">9</span>       <span class="m">0</span>.75    1F      1F
<span class="m">11</span>      <span class="m">1</span>       <span class="m">10</span>      <span class="m">0</span>.5263157894736842      1F      1F
<span class="m">13</span>      <span class="m">8</span>       <span class="m">1</span>       <span class="m">0</span>.02127659574468085     1F      1F
<span class="m">2</span>       <span class="m">9</span>       <span class="m">12</span>      <span class="m">0</span>.3333333333333333      1F      1F
</pre></div>
</div>
</section>
<section id="crispr-summary-table-csv">
<span id="blast-crispr-sum-csv"></span><h3>CRISPR_summary_table.csv<a class="headerlink" href="#crispr-summary-table-csv" title="Permalink to this headline"></a></h3>
<p><strong>Summary</strong></p>
<p>Summary of CRISPR arrays found in each assembly with information about each array. This file is designed to be read into Microsoft Excel or a similar program to view.</p>
<p><strong>Format</strong></p>
<p>comma-delimited (csv) table</p>
<p>Columns:</p>
<ol class="arabic simple">
<li><p>Sequence_ID: Name of assembly (extracted from input file name)</p></li>
<li><p>Has_CRISPR: Boolean whether and CRISPR arrays were found</p></li>
<li><p>Array_count: Number of CRISPR arrays found. No further columns are populated if no arrays were found.</p></li>
<li><p>Spacers: List of spacer sequences found in each array</p></li>
<li><p>Spacer_IDs: List of spacer IDs found in each array</p></li>
<li><p>Array_IDs: List of array IDs</p></li>
<li><p>Array_locations: List of array locations (contig name, start, stop)</p></li>
<li><p>Repeat_sequences: Sequence of the most common repeat in each array</p></li>
<li><p>Array_CRISPR_types: Most similar repeat type found</p></li>
</ol>
<p>In columns 4-9, arrays are numbered according to the order in which they were found in the input assembly file. These numbers correspond between columns in a given row such that the spacer IDs for array 1 correspond to the spacer sequences of array 1 etc.</p>
<p><strong>Example</strong></p>
<img alt="../_images/cr_sum_tab.png" src="../_images/cr_sum_tab.png" />
</section>
<section id="crispr-summary-table-txt">
<span id="blast-crispr-sum-txt"></span><h3>CRISPR_summary_table.txt<a class="headerlink" href="#crispr-summary-table-txt" title="Permalink to this headline"></a></h3>
<p><strong>Summary</strong></p>
<p>Summary of CRISPR arrays found in each assembly with information about each array. This file is easier to interact with programatically.</p>
<p><strong>Format</strong></p>
<p>Tab-delimited table with “|” (pipe)-delimited lists of arrays in columns 4-9 within each array, elements are space-delimited.</p>
<p><strong>Example</strong></p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Sequence_ID     Has_CRISPR      Array_count     Spacers Spacer_IDs      Array_IDs       Array_locations Repeat_sequences        Array_CRISPR_types
Assembly1       True    <span class="m">3</span>       TAGCTGATCAGCAGGCCGACAGTCAGGCCTGC TACCCGAATACGACTTGCGCGAGGAAGACGGT AGCATCGCATCAAATCGTGCAGAACACGATAA TGGTCGAGCAGTTCGGCAAAGGGGCCGTGGTT TTCACCTGGTCGCCGGCCAGGCTGATCACTGC TACAAGGTCATGGCGCTCGGCAACGTGGTGGAA GCTGTGCGTCGCCGTGGTCTGACGGTCGAATC AGCAGATACCCGAACCACTGGAGGTACATGCA TTCATCAGGATGCCGCCAAGGGTCCGCATAAT<span class="p">|</span>AGGTCGAGGTGGGCTCGGCGGCGATGATCGAT GGTACGTGGTTTCGACCAACAGCACTGCCCAA TAAAGGAGATTGCCATGCTGATCAAACTTCCC GTCAGGGTCGTGCATGACTCCGATGTGGTGGC CGTCCAGAACGTCACACGCTCGCCGTCGATGT AACCGGAGCCTTCGGGCCGCGTTGGGATCCAC TTGACTGCTGGGGCCTGACGCTCATCGCGCGG GCGACCCTGGCCAGGGCGGCGTCGCGCTCTGC TTGAGCACAACCGGCTGAGCCAGCTGGTTGTC<span class="p">|</span>CAGCAGCGGCTCCAGGAAGAGGGGCGCTGCCT AAGAGTCGCGGCGACAACTACCAGACGTCCGC GTATGGCTCTCTCCATTGGGGTGGCGATACTC GATCTGGGGCGGCATCATCACAGCAGAATCTA ACAACATCAATCGCCTGATGCTGGGGCACCTG AGCTTCGGCACCCTGATGCGCGCCGTCGAGGG AATGCGGTCCTGCGCATCCGAACTGGTAAGTG GACCCCCGGAGGACCAACCGTGGACAACGACA TCCTTCGGCTCCGCCGGCCGGATCGCTGCAT GTCGCGAAGTTCATAAGCGGGCTTAGGGCGA      1F_156 1F_19 1F_46 1F_123 1F_59 1F_64 1F_34 1F_93 1F_33<span class="p">|</span>1F_99 1F_1 1F_45 1F_83 1F_124 1F_126 1F_30 1F_39 1F_49<span class="p">|</span>1F_134 1F_81 1F_55 1F_84 1F_16 1F_5 1F_51 1F_100 1F_106 1F_145 <span class="m">6</span><span class="p">|</span><span class="m">7</span><span class="p">|</span><span class="m">11</span>    Assembly1_contig2 <span class="m">209013</span>-208445<span class="p">|</span>Assembly1_contig4 <span class="m">19992</span>-20559<span class="p">|</span>Assembly1_contig4 <span class="m">30050</span>-29425 GTTCACTGCCGTATAGGCAGCTAAGAAA<span class="p">|</span>GTTCACTGCCGTGTAGGCAGCTAAGAAA<span class="p">|</span>GTTCACTGCCGTATAGGCAGCTAAGAAA      1F<span class="p">|</span>1F<span class="p">|</span>1F
</pre></div>
</div>
</section>
<section id="spacer-cluster-members-txt">
<span id="blast-spacer-cluster-reps"></span><h3>Spacer_cluster_members.txt<a class="headerlink" href="#spacer-cluster-members-txt" title="Permalink to this headline"></a></h3>
<p><strong>Summary</strong></p>
<p>When running <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> with <code class="docutils literal notranslate"><span class="pre">-s</span></code> to cluster similar spacers, this file is produced to provide details of which spacers were identified as similar to one another.</p>
<p><strong>Format</strong></p>
<p>Tab-delimited table with two columns. Each line represents a distinc cluster of spacers. Column 1 is the ID of the spacer chosen as the representative of the cluster. The ID (or its corresponding sequence - see CRISPR_spacers.fna) is used to represent all cluster members in any files in which they are described. Column 2 is a space-delimited list of the sequences of spacers that are members of the cluster (not including the sequence of the spacer chosen as the representative.)</p>
<p><strong>Example</strong></p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>CRtype_1        GCCCAGGCACGTTTGCTCGCGCTTTGATCTCA
CRtype_13       TGTCCCGAAGTTCATAAGCGGGCTTCGGGCGA GTCGCGAAGTTCATAAGCGGGCTTCGGGCGA
CRtype_42       AGCCGATGGCCCGCAGTAGTACCCCGATCAGT
</pre></div>
</div>
</section>
</section>
<section id="advanced-usage">
<span id="blast-advanced"></span><h2>Advanced Usage<a class="headerlink" href="#advanced-usage" title="Permalink to this headline"></a></h2>
<section id="associating-arrays-from-different-contigs">
<h3>Associating arrays from different contigs<a class="headerlink" href="#associating-arrays-from-different-contigs" title="Permalink to this headline"></a></h3>
<p>It is highly recommended to also provide information about how to relate sequences in the BLASTdb together. Without this information, each array will only be associated to the sequence ID of the contig in which it was found, rather than all arrays in a given assembly being associated with that assembly. You can provide this information in two forms: a regex that describes identifying information present in all sequences from the same assembly, or a file containing a list of identifiers that are present in your sequence headers. An example of how this regex and file may look are provided in .</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Either:

cctk blast -d &lt;BLASTdb&gt; -r &lt;repeats file&gt; -o &lt;output dir&gt; -a &lt;file containing IDs&gt;

OR:

cctk blast -d &lt;BLASTdb&gt; -r &lt;repeats file&gt; -o &lt;output dir&gt; -p &lt;regex pattern&gt;
</pre></div>
</div>
<section id="example-regex-or-id-file">
<h4>Example regex or ID file<a class="headerlink" href="#example-regex-or-id-file" title="Permalink to this headline"></a></h4>
<p>In this example, we have assemblies made from reads retrieved from the SRA database. Our assembly files are names with the accession number of the reads and we have added the file name to each contig’s fasta header as described in the <a class="reference internal" href="#blast-before-you-run"><span class="std std-ref">Before you run</span></a> section. Our file names look something like this:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>SRR146516.fasta
SRR56754356.fasta
...
</pre></div>
</div>
<p>These IDs can be described using the following regex. Either “SRR[0-9]+” or “SRRd+”. You can test your regex using the function <code class="docutils literal notranslate"><span class="pre">grep</span> <span class="pre">-haoP</span></code> or <code class="docutils literal notranslate"><span class="pre">grep</span> <span class="pre">-haoE</span></code> with your regex pattern against the .nhr file of your blastdb. E.g.:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>grep -haoP <span class="s2">&quot;SRR\d+&quot;</span> &lt;BLASTdb&gt; <span class="p">|</span> sort <span class="p">|</span> uniq
</pre></div>
</div>
<p>If that command returns the full list of accession numbers that you expected to see then it will work as an input to <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code>.</p>
<p>An alternative to providing a regex is to provide a list of identifiers to use. In the case of the example files described abode this file would look like this:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>SRR146516
SRR56754356
...
</pre></div>
</div>
<p>This can be easily produced with the following command:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ls <span class="p">|</span> sed <span class="s1">&#39;s/.fasta//&#39;</span> &gt; ID_file.txt
</pre></div>
</div>
</section>
</section>
</section>
<section id="limitations-and-considerations">
<span id="blast-limitations"></span><h2>Limitations and considerations<a class="headerlink" href="#limitations-and-considerations" title="Permalink to this headline"></a></h2>
<p>As described in the <a class="reference internal" href="#blast-intro"><span class="std std-ref">Introduction</span></a> section, <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> is built around the <code class="docutils literal notranslate"><span class="pre">blastn</span></code> program of the NCBI BLAST+ suite. You can therefore expect many of the same behaviours.</p>
<p><code class="docutils literal notranslate"><span class="pre">BLASTn</span></code> and <code class="docutils literal notranslate"><span class="pre">minced</span></code> have different tolerance of mutations in sequences when identifying repeats. See <a class="reference internal" href="tutorial.html#minced-blast-comp"><span class="std std-ref">cctk minced vs cctk blast output comparison</span></a> for an example of how these defferences result in the identification of differing arrays.</p>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> searches for repeats based on sequence identity. It is therefore only useful if you already know the CRISPR repeat sequence you are looking for. <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> is a better choice if you are exploring sequences that you expect to contain CRISPRs of unknown types.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="CRISPRtree.html" class="btn btn-neutral float-left" title="CRISPRtree" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="constrain.html" class="btn btn-neutral float-right" title="constrain" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Alan Collins.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>