<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>spacerblast &mdash; cctk 0.3.3 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tutorial" href="tutorial.html" />
    <link rel="prev" title="network" href="network.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> cctk
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="CRISPRdiff.html">CRISPRdiff</a></li>
<li class="toctree-l1"><a class="reference internal" href="CRISPRtree.html">CRISPRtree</a></li>
<li class="toctree-l1"><a class="reference internal" href="blast.html">blast</a></li>
<li class="toctree-l1"><a class="reference internal" href="constrain.html">constrain</a></li>
<li class="toctree-l1"><a class="reference internal" href="evolve.html">evolve</a></li>
<li class="toctree-l1"><a class="reference internal" href="minced.html">minced</a></li>
<li class="toctree-l1"><a class="reference internal" href="network.html">network</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">spacerblast</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#before-you-run">Before you run</a></li>
<li class="toctree-l2"><a class="reference internal" href="#basic-usage">Basic Usage</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#simple-blast-with-match-extension">Simple BLAST with match extension</a></li>
<li class="toctree-l3"><a class="reference internal" href="#checking-flanking-sequence-for-pam">Checking flanking sequence for PAM</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#output-files">Output files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#o-main-output-file-hits-with-pams"><code class="docutils literal notranslate"><span class="pre">-o</span></code> Main output file / hits with PAMs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#q-hits-without-pams"><code class="docutils literal notranslate"><span class="pre">-q</span></code> Hits without PAMs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#advanced-usage">Advanced Usage</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#control-amount-and-location-of-flanking-sequence-retrieved">Control amount and location of flanking sequence retrieved</a></li>
<li class="toctree-l3"><a class="reference internal" href="#specify-a-pam-using-a-regex">Specify a PAM using a regex</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mask-regions-of-sequences-in-you-blastdb">Mask regions of sequences in you blastdb</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">cctk</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>spacerblast</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/content/spacerblast.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="spacerblast">
<h1>spacerblast<a class="headerlink" href="#spacerblast" title="Permalink to this headline"></a></h1>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">spacerblast</span></code> is essentially a wrapper script for <code class="docutils literal notranslate"><span class="pre">blastn</span></code> that performs additional functions that are useful when searching for the targets of CRISPR spacers. The additional functions performed by <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">spacerblast</span></code> are:</p>
<ol class="arabic simple">
<li><p>Extends sequence matches to cover the entire length of the query sequence. For short query sequences, a small number of mismatches can result in the BLAST algorithm not extending the match through the mismatches even if additional sequence identity would be found. Users can specify a percent identity requirement that takes into account the full sequence length.</p></li>
<li><p>Retrieves flanking sequence either side (or both sides) of the match location in the subject sequence.</p></li>
<li><p>Identifies the presence of user-defined sequence motifs in sequence flanking the match and sorts blast hits based on the presence or absence of the motif.</p></li>
</ol>
<ol class="arabic simple" start="3">
<li><p>Masks user-specified regions in the subject sequence and ignores blast hits in those regions.</p></li>
</ol>
</section>
<section id="before-you-run">
<h2>Before you run<a class="headerlink" href="#before-you-run" title="Permalink to this headline"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">spacerblast</span></code> requires you to provide the sequences you wish to search in the form of a blast database. This can be acheived simply using the <code class="docutils literal notranslate"><span class="pre">makeblastdb</span></code> command from NCBI BLAST+ which is included in the conda installation of CCTK. However, before making your blastdb, please confirm that your sequences meet the following requirements:</p>
<ol class="arabic simple">
<li><p>No pipe symbols (“|”) in any of your fasta headers.</p></li>
<li><p>None of the fasta headers in the sequences are the same.</p></li>
</ol>
<p>Making a blastdb from a directory containing your sequences to search can be acheived in two steps (if you are only searching a single sequence file skip the first step). Assuming your sequences are in a directory called sequences/:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>cat sequences/* &gt; all_seqs.fna
makeblastdb -in all_seqs.fna -out seq_blastdb -dbtype nucl -parse_seqids
</pre></div>
</div>
<p><strong>N.B.</strong> It is essential to include the <code class="docutils literal notranslate"><span class="pre">-parse_seqids</span></code> option when creating the blastdb. You can be check if your blastdb was made using <code class="docutils literal notranslate"><span class="pre">-parse_seqids</span></code> by looking for files ending in .nog and .nos. If those files exist associated with your blastdb then it was.</p>
</section>
<section id="basic-usage">
<h2>Basic Usage<a class="headerlink" href="#basic-usage" title="Permalink to this headline"></a></h2>
<section id="simple-blast-with-match-extension">
<h3>Simple BLAST with match extension<a class="headerlink" href="#simple-blast-with-match-extension" title="Permalink to this headline"></a></h3>
<p>The most basic usage requires only 2 inputs: the blastdb name (i.e. without file extensions) to search and the query sequences.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>cctk spacerblast -d &lt;blastdb path&gt; -s &lt;query sequence file&gt;
</pre></div>
</div>
<p>N.B. If you are on a computer with multiple threads, you can speed up <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">spacerblast</span></code> by specifying a number of threads to use with the <code class="docutils literal notranslate"><span class="pre">-t</span></code> option.</p>
<p>Not specifying an output file results in the output being directed to the stdout for easy downstream filtering with command-line tools. The above command results in an output like the example below:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Spacer_ID       Target_contig   Protospacer_start       Protospacer_end Percent_identity        mismatches      protospacer_sequence    target_strand
1F_1    Assembly1_contig1       <span class="m">66234</span>   <span class="m">66265</span>   <span class="m">100</span>.0   <span class="m">0</span>       GGTACGTGGTTTCGACCAACAGCACTGCCCAA        minus
1F_1    Assembly5_contig2       <span class="m">91689</span>   <span class="m">91720</span>   <span class="m">78</span>.125  <span class="m">7</span>       GGTACGTGGTTTCGACAAGCAACGTGGCCCAG        plus
1F_1    Assembly6_contig1       <span class="m">15636</span>   <span class="m">15667</span>   <span class="m">59</span>.375  <span class="m">13</span>      TGGTGCACGTTTCGACCAACAGCCTAGCGCCC        plus
</pre></div>
</div>
</section>
<section id="checking-flanking-sequence-for-pam">
<h3>Checking flanking sequence for PAM<a class="headerlink" href="#checking-flanking-sequence-for-pam" title="Permalink to this headline"></a></h3>
<p>You can also retrieve just the hits that have an adjacent PAM sequence. For example, to check for the <em>Pseudomonas aeruginosa</em> type I-F PAM, GG, which occurs 5’ (or upstream) of the protospacer:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>cctk spacerblast -d &lt;blastdb path&gt; -s &lt;query sequence file&gt; -P GG -l up
</pre></div>
</div>
<p>You can specify your PAM in this way using any characters in the <a class="reference external" href="https://www.bioinformatics.org/sms/iupac.html">IUPAC nucleotide alphabet</a>.</p>
<p>The above command returns something like the following output:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Your specified PAM is at least <span class="m">2</span> bases, but you only requested <span class="m">0</span> upstream bases. <span class="m">2</span> bases will now be retrieved on the upstream side.
Spacer_ID       Target_contig   Protospacer_start       Protospacer_end Percent_identity        mismatches      protospacer_sequence    upstream_bases  target_strand
1F_1    Assembly1_contig1       <span class="m">66234</span>   <span class="m">66265</span>   <span class="m">100</span>.0   <span class="m">0</span>       GGTACGTGGTTTCGACCAACAGCACTGCCCAA        GG      minus
1F_1    Assembly6_contig1       <span class="m">15636</span>   <span class="m">15667</span>   <span class="m">59</span>.375  <span class="m">13</span>      TGGTGCACGTTTCGACCAACAGCCTAGCGCCC        GG      plus
</pre></div>
</div>
<p>There are two differences to note here between the first output and the output when specifying a PAM sequence:</p>
<ol class="arabic simple">
<li><p>The output includes an additional column “upstream_bases” which contains the 2 bases flanking the protospacer on the 5’ side. If the PAM had been specified as being on the other side (i.e. <code class="docutils literal notranslate"><span class="pre">-l</span> <span class="pre">down</span></code>) then this column would be “downstream_bases”.</p></li>
<li><p>A warning message was written to stderr stating that our PAM is two bases long while we didn’t ask for any flanking bases to be checked. This message is intended to make clear why a different number of bases are returned if you specifically request fewer bases than the provided PAM requires. <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">spacerblast</span></code> will automatically determine the shortes length of sequence that your PAM could match and will return at least that much sequence.</p></li>
</ol>
</section>
</section>
<section id="output-files">
<h2>Output files<a class="headerlink" href="#output-files" title="Permalink to this headline"></a></h2>
<p>The default behaviour of <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">spacerblast</span></code> is to direct ouputs to the stdout and information and error messages to stderr. However, two output files can be produced if requested by the user using the <code class="docutils literal notranslate"><span class="pre">-o</span></code> and <code class="docutils literal notranslate"><span class="pre">-q</span></code> options.</p>
<section id="o-main-output-file-hits-with-pams">
<h3><code class="docutils literal notranslate"><span class="pre">-o</span></code> Main output file / hits with PAMs<a class="headerlink" href="#o-main-output-file-hits-with-pams" title="Permalink to this headline"></a></h3>
<p>This option directs any output that would have been sent to stdout to the specified file instead. You can name this file and specify its location by providing the path to a file (i.e. <code class="docutils literal notranslate"><span class="pre">-o</span> <span class="pre">&lt;path</span> <span class="pre">to</span> <span class="pre">file&gt;</span></code>)</p>
<p>If no PAM information is provided then this output file contains all BLAST hits that meet the percent identity and evalue thresholds. If PAM information is provided, this file will contain just the hits that were found to have an adjacent PAM.</p>
</section>
<section id="q-hits-without-pams">
<h3><code class="docutils literal notranslate"><span class="pre">-q</span></code> Hits without PAMs<a class="headerlink" href="#q-hits-without-pams" title="Permalink to this headline"></a></h3>
<p>This file will only be generated if PAM information is provided. You can name this file and specify its location by providing the path to a file (i.e. <code class="docutils literal notranslate"><span class="pre">-q</span> <span class="pre">&lt;path</span> <span class="pre">to</span> <span class="pre">file&gt;</span></code>).</p>
<p>If PAM information is provided, this file will contain all BLAST hits that were not found to have an adjacent PAM. Only hits that exceed the percent identity and evalue thresholds will be stored.</p>
</section>
</section>
<section id="advanced-usage">
<h2>Advanced Usage<a class="headerlink" href="#advanced-usage" title="Permalink to this headline"></a></h2>
<p>Advanced usage of <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">spacerblast</span></code> is not much more complicated than the basic usage described above. There are three cases in which a more complicated usage is required:</p>
<section id="control-amount-and-location-of-flanking-sequence-retrieved">
<h3>Control amount and location of flanking sequence retrieved<a class="headerlink" href="#control-amount-and-location-of-flanking-sequence-retrieved" title="Permalink to this headline"></a></h3>
<p>The amount of sequence retrieved from each side of BLAST hits can be controlled using command line input with the optione <code class="docutils literal notranslate"><span class="pre">-n</span></code>, <code class="docutils literal notranslate"><span class="pre">-u</span></code>, and <code class="docutils literal notranslate"><span class="pre">-w</span></code>. If also specifying a PAM, at least enough sequence to match the PAM will be retrieved. If you request less sequence than is required to match the provided PAM, the length of sequence retrieved will be adjusted and an informative message will be written to stderr informing you.</p>
<p><code class="docutils literal notranslate"><span class="pre">-n</span></code> can be used to retrieve the same length of flanking sequence on both sides of BLAST hits,</p>
</section>
<section id="specify-a-pam-using-a-regex">
<h3>Specify a PAM using a regex<a class="headerlink" href="#specify-a-pam-using-a-regex" title="Permalink to this headline"></a></h3>
<p>If you would prefer to define your PAM as a regex rather than using IUPAC nucleotide codes, you can do that using the <code class="docutils literal notranslate"><span class="pre">-R</span></code> option. Regex PAM definition is useful when the number of bases is flexible or if you prefer to specify e.g. A, T, or G with “[ATG]” rather than using the IUPAC “D”.</p>
</section>
<section id="mask-regions-of-sequences-in-you-blastdb">
<h3>Mask regions of sequences in you blastdb<a class="headerlink" href="#mask-regions-of-sequences-in-you-blastdb" title="Permalink to this headline"></a></h3>
<p>If you would like to ignore hits in certain regions of your subject sequences you can maks regions by providing a <a class="reference external" href="https://en.wikipedia.org/wiki/BED_(file_format)#Format">BED format</a> file with the <code class="docutils literal notranslate"><span class="pre">-r</span></code> option. Only the first 3 columns of the .bed file will be read so all other columns are optional.</p>
<p>This can be useful when extracting spacers and searching for CRISPR targets in the same set of sequences. It will allow you to ignore hits against CRISPR arrays as each spacer will return a perfect match against its location in the genome in which it was found. Both <a class="reference external" href="blast.html">cctk blast</a> and <a class="reference external" href="minced.html">cctk minced</a> return a .bed file of CRISPR array locations that can be used for this purpose.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="network.html" class="btn btn-neutral float-left" title="network" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorial.html" class="btn btn-neutral float-right" title="Tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Alan Collins.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>