<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Tutorial &mdash; cctk 0.5.3 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="spacerblast" href="spacerblast.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> cctk
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="CRISPRdiff.html">CRISPRdiff</a></li>
<li class="toctree-l1"><a class="reference internal" href="CRISPRtree.html">CRISPRtree</a></li>
<li class="toctree-l1"><a class="reference internal" href="blast.html">blast</a></li>
<li class="toctree-l1"><a class="reference internal" href="constrain.html">constrain</a></li>
<li class="toctree-l1"><a class="reference internal" href="evolve.html">evolve</a></li>
<li class="toctree-l1"><a class="reference internal" href="minced.html">minced</a></li>
<li class="toctree-l1"><a class="reference internal" href="network.html">network</a></li>
<li class="toctree-l1"><a class="reference internal" href="spacerblast.html">spacerblast</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#quick-worked-example">Quick worked example</a></li>
<li class="toctree-l2"><a class="reference internal" href="#detailed-worked-example">Detailed Worked example</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#data-retrieval">Data retrieval</a></li>
<li class="toctree-l3"><a class="reference internal" href="#identifying-crispr-arrays">Identifying CRISPR arrays</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#identifying-crispr-arrays-using-cctk-minced">Identifying CRISPR arrays using <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#identifying-crispr-arrays-using-cctk-blast">Identifying CRISPR arrays using <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#exploring-crispr-array-relationships-using-a-network-representation">Exploring CRISPR array relationships using a network representation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#using-crisprdiff-to-visualize-array-relationships">Using CRISPRdiff to visualize array relationships</a></li>
<li class="toctree-l4"><a class="reference internal" href="#using-crisprtree-to-create-hypotheses-of-array-histories">Using CRISPRtree to create hypotheses of array histories</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cctk-minced-vs-cctk-blast-output-comparison"><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> vs <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> output comparison</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">cctk</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Tutorial</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/content/tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="tutorial">
<h1>Tutorial<a class="headerlink" href="#tutorial" title="Permalink to this headline"></a></h1>
<section id="quick-worked-example">
<h2>Quick worked example<a class="headerlink" href="#quick-worked-example" title="Permalink to this headline"></a></h2>
<p>TO ADD</p>
</section>
<section id="detailed-worked-example">
<h2>Detailed Worked example<a class="headerlink" href="#detailed-worked-example" title="Permalink to this headline"></a></h2>
<section id="data-retrieval">
<h3>Data retrieval<a class="headerlink" href="#data-retrieval" title="Permalink to this headline"></a></h3>
<p>For this tutorial we will work through and analyze part of the dataset used in the CCTK publication, which is itself a subset of the data presented by <a class="reference external" href="https://doi.org/10.1128/mSystems.00075-18">England et al., 2018</a>. The example data used here are hosted on the <a class="reference external" href="https://github.com/Alan-Collins/CRISPR_comparison_toolkit">CCTK github page</a> and can be downloaded and extracted as follows:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>wget https://github.com/Alan-Collins/CRISPR_comparison_toolkit/raw/main/Example_data/Example_data.tar.gz -O - <span class="p">|</span> tar -xz
</pre></div>
</div>
<p>This should download and extract a directory containing 10 <em>Pseudomonas aeruginosa</em> assemblies and some files we will use later. Each assembly filename corresponds to the <a class="reference external" href="https://www.ebi.ac.uk/ena/browser/home">ENA</a> sequence accession for the corresponding sequencing run. These assemblies were produced using <a class="reference external" href="https://github.com/ablab/spades">SPAdes</a>.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ ls -1 Example_assemblies
ERR431075.fasta
ERR431089.fasta
ERR431110.fasta
ERR431125.fasta
ERR431274.fasta
ERR431323.fasta
ERR431333.fasta
ERR431387.fasta
ERR431415.fasta
ERR431427.fasta
</pre></div>
</div>
</section>
<section id="identifying-crispr-arrays">
<h3>Identifying CRISPR arrays<a class="headerlink" href="#identifying-crispr-arrays" title="Permalink to this headline"></a></h3>
<p>If you haven’t already, install CCTK into a <code class="docutils literal notranslate"><span class="pre">conda</span></code> environment</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>conda create -n cctk -c alan-collins cctk
conda activate cctk
</pre></div>
</div>
<p>For this example, we will use both of the CRISPR identification tools included in CCTK to illustrate how they both work with the same input data. However, for simplicity, output from <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> will be used in subsequent analysis sections.</p>
<p>See the <a class="reference internal" href="#minced-blast-comp"><span class="std std-ref">cctk minced vs cctk blast output comparison</span></a> section for a discussion of how the output of <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> and <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> compare to get an idea of why you might choose one over the other.</p>
<section id="identifying-crispr-arrays-using-cctk-minced">
<h4>Identifying CRISPR arrays using <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code><a class="headerlink" href="#identifying-crispr-arrays-using-cctk-minced" title="Permalink to this headline"></a></h4>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> is the easier to use of the two CRISPR identification tools included in CCTK and is a good choice to start with.</p>
<p>First, an important note: For many of the downstream analyses in CCTK, CRISPR array orientation is very important. It is therefore crucial that you know the sequence of CRISPR repeats in your assemblies in the correct orientation (i.e. leader end 5’ of the repeat, trailer end 3’). If you do not yet know the correct orientation of your repeats, you can proceed with these analyses as <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> can identify CRISPR arrays with unknown repeats and will consistently orient them using a built-in database of repeats. However, you will want to come back and correctly orient your arrays before using the tools <code class="docutils literal notranslate"><span class="pre">CRISPRtree</span></code> or <code class="docutils literal notranslate"><span class="pre">constrain</span></code>.</p>
<section id="setting-up-and-running-minced">
<h5>Setting up and running MinCED<a class="headerlink" href="#setting-up-and-running-minced" title="Permalink to this headline"></a></h5>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> uses <a class="reference external" href="https://github.com/ctSkennerton/minced">MinCED</a> to identify CRISPR arrays and then processes the output files produced by MinCED. <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> runs MinCED on all of the files in the input directory. In our case the directory is “Example_assemblies/”. Before we run it, we’ll make an directory to contain the output of <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code>, then we can run it. While <cite>cctk minced</cite> includes the three <em>P. aeruginosa</em> CRISPR types in its built-in database of repeats, we will provide them here to illustrate that step. They were included in the data you downloaded above: example_repeats.fna.</p>
<p>First let’s just make a directory for the output and run the MinCED step. We could run everything in one step using <code class="docutils literal notranslate"><span class="pre">-m</span> <span class="pre">-p</span></code>, but we will split it up here to go through the process in more detail.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make output dir</span>
mkdir Minced_CRISPRs/

<span class="c1"># Run cctk minced</span>
cctk minced -i Example_assemblies/ -o Minced_CRISPRs/ -m
</pre></div>
</div>
</section>
<section id="exploring-minced-output-before-processing-using-cctk-minced">
<h5>Exploring MinCED output before processing using <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code><a class="headerlink" href="#exploring-minced-output-before-processing-using-cctk-minced" title="Permalink to this headline"></a></h5>
<p>After a few seconds, you will see that a new directory has been created in the Minced_CRISPRs/ directory.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ ls Minced_CRISPRs/
MINCED_OUT
</pre></div>
</div>
<p>This new directory contains all of the output files produced by MinCED. Each of these files is named using the base filename (i.e. excluding extension) of the input files, with “_minced_out.txt” added.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ ls -1 Minced_CRISPRs/MINCED_OUT/
ERR431075_minced_out.txt
ERR431089_minced_out.txt
ERR431110_minced_out.txt
ERR431125_minced_out.txt
ERR431274_minced_out.txt
ERR431323_minced_out.txt
ERR431333_minced_out.txt
ERR431387_minced_out.txt
ERR431415_minced_out.txt
ERR431427_minced_out.txt
</pre></div>
</div>
<p>These output files are the unaltered output produced by MinCED and are where <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> gets all the information it uses (except reference repeat sequences provided by the user).</p>
<p>If you are working with assemblies with unknown repeats, you can easily extract a list of the repeats found by MinCED and count the number of times each repeat sequence occurs by pulling out nucleotide sequence from the repeat column in the MinCED output files. For the assemblies analyzed here, the following repeats were found:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>    $ cut -f3 Minced_CRISPRs/MINCED_OUT/* <span class="p">|</span> grep -Po <span class="s2">&quot;[ATCG]{15,100}&quot;</span> <span class="p">|</span> sort <span class="p">|</span> uniq -c <span class="p">|</span> sort -k1 -nr
<span class="m">127</span> GTTCACTGCCGTATAGGCAGCTAAGAAA
<span class="m">123</span> TTTCTTAGCTGCCTACACGGCAGTGAAC
 <span class="m">99</span> TTTCTTAGCTGCCTATACGGCAGTGAAC
 <span class="m">26</span> GTTCACTGCCGTGTAGGCAGCTAAGAAA
 <span class="m">13</span> CGGTTCATCCCCACGCATGTGGGGAACAC
  <span class="m">8</span> CGGTTCATCCCCACACCCGTGGGGAACAC
  <span class="m">5</span> TTTCTTAGCTGCCTGTACGGCAGTGAAC
  <span class="m">5</span> TTTCACTGCCACATAGGTCGTCAAGAAA
  <span class="m">5</span> GTTCACTGCCGTACAGGCAGCTAAGAAA
  <span class="m">3</span> TTTCTTAGCTACCTATACGGCAGTGAAC
  <span class="m">3</span> GTTCACTGCCGTATAGGCAGCTAAAAAA
  <span class="m">2</span> TTTCTTGACGACCTATGTGGCAGTGAAA
  <span class="m">1</span> TTTTTTAGCTGCCTATACGGCAGTGAAC
  <span class="m">1</span> TTTCTTAGCTGCCTATACGGCAGTGGAC
  <span class="m">1</span> TTTCTTAGCGGCCTACACGGCAGTGAAC
  <span class="m">1</span> TTTCTTAACTGCCTATACGGCAGTGAAC
  <span class="m">1</span> TGGTTCATCCCCACGCATGTGGGGAACAC
  <span class="m">1</span> GTTCACTGCCGTGTAGGCCGCTAAGAAA
  <span class="m">1</span> GTTCACTGCCGTATCGGCAGCTAAGAAA
  <span class="m">1</span> GTTCACTGCCGTATAGGTAGCTAAGAAA
  <span class="m">1</span> GTTCACTGCCGTATAGGCAGCAAAAAAT
  <span class="m">1</span> GTTCACTGCCGTATAGGCAGCAAAAAAA
  <span class="m">1</span> GTCCACTGCCGTATAGGCAGCTAAGAAA
  <span class="m">1</span> CAGGTTTATCCCACGCCTGTGGGGGAACA
  <span class="m">1</span> CAGCTCATCCCCACGCCCGTGGGGAACAC
  <span class="m">1</span> ATTTTTTGCTGCCTATACGGCAGTGAAC
</pre></div>
</div>
<p>In the output returned by this command, the second column is the nucleotide sequence of each repeat found by MinCED, while the first column is the number of instances of that repeat that were found. Note that the first 2 lines are highly similar sequences that are in different orientations to one another. In fact the first 4 lines are minor variations on the same sequence. This highlights a weakness of MinCED that it is important to be aware of: it does not check the orientation of repeats so there is a roughly 50% chance that the repeat will be correctly oriented in any given sequence.</p>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> attempts to consistently orient CRISPR arrays by comparing each query repeat to a built-in database of repeat sequences, or a user-provided database of repeats (reverse complement is also checked). If the query repeat matches best in the reverse complement then the entire corresponding array is reversed. Even if the specific repeat present in your assemblies is not in the built-in database, there is a good chance that all instances of that repeat (even with some SNPs) will match the same reference repeat, resulting in consistent (even if incorrect) orientation even for unknown repeats. We will discuss below what happens when repeats don’t match well with any of the reference repeats used by <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code></p>
</section>
<section id="processing-minced-output-files">
<h5>Processing MinCED output files<a class="headerlink" href="#processing-minced-output-files" title="Permalink to this headline"></a></h5>
<p>Now that we have briefly explored the output produced by MinCED, we will use <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> to process those output files. As we have already run MinCED, we don’t need <code class="docutils literal notranslate"><span class="pre">-m</span></code> or to point <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> to our assemblies with <code class="docutils literal notranslate"><span class="pre">-i</span></code>.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ cctk minced -o Minced_CRISPRs/ -p
Total unique spacers: <span class="m">244</span>
Total unique arrays: <span class="m">21</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> outputs a summary of the number of unique spacers and CRISPR arrays identified to stderr. Output files produced here are described in the <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> <a class="reference internal" href="minced.html#minced-output"><span class="std std-ref">Output files</span></a> section.</p>
<p>As we did not provide a file containing FASTA format CRISPR repeats, <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> uses the built-in database of CRISPR repeats (currently subtypes I-A, I-C, I-E, and I-F). The built-in database is fine for the assemblies used here as <em>P. aeruginosa</em> encodes CRISPR subtypes I-C, I-E, and I-F.</p>
<p>Note that each spacer has its predicted CRISPR subtype included in its FASTA header and this header is used as an identifier for each spacer in all other output files.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ head -4 Minced_CRISPRs/PROCESSED/CRISPR_spacers.fna
&gt;1F_1
GCCCCAGATCAGCGTTGCGCCGCACGCGATGA
&gt;1F_2
AGTAGCCGTCGGCGTTGTGACAAAGCCATTCC
</pre></div>
</div>
</section>
<section id="specifying-expected-repeat-sequences">
<h5>Specifying expected repeat sequences<a class="headerlink" href="#specifying-expected-repeat-sequences" title="Permalink to this headline"></a></h5>
<p>Now let’s process the MinCED output using the example_repeats.fna file that we downloaded earlier. This file contains sequences for the three CRISPR subtypes found in <em>P. aeruginosa</em>, but the FASTA headers for those sequences differ from the IDs used in the built-in CRISPR database (e.g. “1F” above vs “I-F” in the example file):</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ cat example_repeats.fna
&gt;I-C
GTCGCGCCCCGCACGGGCGCGTGGATTGAAAC
&gt;I-E
GTGTTCCCCACGGGTGTGGGGATGAACCG
&gt;I-F
GTTCACTGCCGTGTAGGCAGCTAAGAAA
</pre></div>
</div>
<p>We run the processing steps again with these repeat sequences by providing them using <code class="docutils literal notranslate"><span class="pre">-r</span></code>. Notice that providing repeat sequences does not change the number of spacers or arrays found. This is only used to decide which direction each array should be and to assign names to spacers.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ cctk minced -o Minced_CRISPRs/ -p -r example_repeats.fna
Total unique spacers: <span class="m">244</span>
Total unique arrays: <span class="m">21</span>
</pre></div>
</div>
<p>Rerunning processing steps will overwrite the existing files in the Minced_CRISPRs/PROCESSED/ directory. If you ever want to preserve the output from a processing run, you will need to rename the PROCESSED folder to something else to prevent the data being overwritten.</p>
<p>When repeats are provided as an input file, the built-in database of CRISPR repeats is not used. All identified CRISPR spacers will therefore be assigned subtypes with the IDs present in the provided file.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ head -4 Minced_CRISPRs/PROCESSED/CRISPR_spacers.fna
&gt;I-F_1
TACTGGAGGTGTCGGAGTTCGAGCCGAGCCAG
&gt;I-F_2
GAAAAGAGTTGACTGCACAGTGGGCATCACCT
</pre></div>
</div>
<p>What about if the repeats in the built-in database or in the file provided using <code class="docutils literal notranslate"><span class="pre">-r</span></code> are not the repeats present in our assemblies? Let’s see.</p>
<p>The example assemblies we are working with here have mostly subtype I-F arrays and a few I-E arrays. Let’s see what happens if we use a repeats file containing only the I-E repeat:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use grep to find the line containing &quot;I-E&quot; and return that line and the line after</span>
$ grep -A1 <span class="s2">&quot;I-E&quot;</span> example_repeats.fna &gt; 1E_repeat.fna
$ cat 1E_repeat.fna
&gt;I-E
GTGTTCCCCACGGGTGTGGGGATGAACCG

<span class="c1"># Now use the new 1E_repeat.fna file as input to cctk minced with -r</span>
$ cctk minced -o Minced_CRISPRs/ -p -r 1E_repeat.fna
Total unique spacers: <span class="m">244</span>
Total unique arrays: <span class="m">21</span>
</pre></div>
</div>
<p>Now if we have a look at our CRISPR spacers we will see that the first couple of spacers in our file, which were previously identified as subtype I-F, have a different header</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ head -4 Minced_CRISPRs/PROCESSED/CRISPR_spacers.fna
&gt;unknown_CRISPR_type<span class="o">(</span>I-E<span class="o">)</span>_1
AACGTCTGCACATTATGCGAAGCCTAGGGTGT
&gt;unknown_CRISPR_type<span class="o">(</span>I-E<span class="o">)</span>_2
GAGGTGGCCATCTGTCGCCCTGGTGATCCGGA
</pre></div>
</div>
<p>When <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> finds that a repeat differs at more than 5 bases from any repeat in the database being used, that repeat is classified as unknown. When a repeat is classified as unknown, the most similar repeat subtype is included in parentheses in the ID for all spacers in the arrays with that repeat. As we only provided a repeat associated with subtype I-E, that repeat is the best matching and is therefore included in the header for all spacers.</p>
</section>
<section id="grouping-very-similar-spacers-based-on-snps">
<h5>Grouping very similar spacers based on SNPs<a class="headerlink" href="#grouping-very-similar-spacers-based-on-snps" title="Permalink to this headline"></a></h5>
<p>Sometimes, due to the acquisiton of mutations, or sequencing errors, CRISPR spacers may be identified that differ only by one or two bases. If you wish spacers that differ by a small amount to be considered the same by <code class="docutils literal notranslate"><span class="pre">cctk</span></code> tools, then you can use <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> to identify groups of highly similar spacers and to assign them the same ID. This is done by using <code class="docutils literal notranslate"><span class="pre">-s</span></code> and providing an integer threshold defining the number of SNPs that can exist between to spacers for those spacers to be considered the same. The deafault behaviour of <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> is to consider spacers that differ by a single base to be different.</p>
<p>This process is performed during the processing steps performed by <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> and so we do not need to run MinCED again. Our command is therefore very similar to those above:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1"># Consider spacers with 2 or fewer SNPs to be the same</span>

$ cctk minced -o Minced_CRISPRs/ -p -s <span class="m">2</span>
Total unique spacers: <span class="m">240</span>
Total unique arrays: <span class="m">21</span>
</pre></div>
</div>
<p>Note that the number of unique spacers identified is now 240 instead of 244.</p>
<p>The spacers that have now been reduced to a single representetive are described in an output file that was not produced by previous runs: <a class="reference internal" href="minced.html#spacer-cluster-reps"><span class="std std-ref">Spacer_cluster_members.txt</span></a></p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ cat Minced_CRISPRs/PROCESSED/Spacer_cluster_members.txt
1F_12   GCCCAGGCACGTTTGCTCGCGCTTTGATCTCA
1F_32   TGTCCCGAAGTTCATAAGCGGGCTTCGGGCGA GTCGCGAAGTTCATAAGCGGGCTTCGGGCGA
1F_33   AGCCGATGGCCCGCAGTAGTACCCCGATCAGT
</pre></div>
</div>
</section>
</section>
<section id="identifying-crispr-arrays-using-cctk-blast">
<h4>Identifying CRISPR arrays using <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code><a class="headerlink" href="#identifying-crispr-arrays-using-cctk-blast" title="Permalink to this headline"></a></h4>
<section id="setting-up">
<h5>Setting up<a class="headerlink" href="#setting-up" title="Permalink to this headline"></a></h5>
<p>Before we can run <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> we must first perform a few steps to prepare our sequences. <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> requires that we provide out input in the form of a blast database.</p>
<p>As described in the <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> <a class="reference internal" href="blast.html#blast-before-you-run"><span class="std std-ref">Before you run</span></a> section, there are several requirements that your sequences must satisfy:</p>
<ol class="arabic simple">
<li><p>No pipe symbols (“|”) in any of your fasta headers.</p></li>
<li><p>None of the fasta headers in the sequences are the same.</p></li>
<li><p>If your sequences are broken up into multiple contigs, ensure that each fasta header contains an identifier that can be used to associate the sequences.</p></li>
</ol>
<p>The example sequences we are working with here were assembled using Spades, which produces assemblies with contig headers that can not be distinguished between files:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>head -1 Example_assemblies/* <span class="p">|</span> head -5
<span class="o">==</span>&gt; Example_assemblies/ERR431075.fasta &lt;<span class="o">==</span>
&gt;NODE_1_length_486033_cov_46.527666

<span class="o">==</span>&gt; Example_assemblies/ERR431089.fasta &lt;<span class="o">==</span>
&gt;NODE_1_length_794353_cov_41.111729
</pre></div>
</div>
<p>If we were to combine these sequences into a single blast database, it would be laborious to later figure out which sequences came from which files. Instead, as each filename contains identifying information (the ERR accession number), we will add that accession to each fasta header in each file. This modification can be acheived with the following bash commands:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> file <span class="k">in</span> Example_assemblies/*<span class="p">;</span> <span class="k">do</span> <span class="nv">id</span><span class="o">=</span><span class="si">${</span><span class="nv">file</span><span class="p">%.*</span><span class="si">}</span><span class="p">;</span> <span class="nv">id</span><span class="o">=</span><span class="si">${</span><span class="nv">id</span><span class="p">##*/</span><span class="si">}</span><span class="p">;</span> sed -i <span class="s2">&quot;s/&gt;/&gt;</span><span class="si">${</span><span class="nv">id</span><span class="si">}</span><span class="s2">_/&quot;</span> <span class="nv">$file</span><span class="p">;</span> <span class="k">done</span>
</pre></div>
</div>
<p>Now all of the fasta headers in our assembly files can easily be related back to the assembly to which they belong.</p>
<p>Now let’s make a directory to contain our blastdb, combine our sequences, and make the bastabase:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>mkdir Blastdb
cat Example_assemblies/* &gt; all_assemblies.fna
makeblastdb -in all_assemblies.fna -out Blastdb/assembly_db -dbtype nucl -parse_seqids
</pre></div>
</div>
<p>We are now ready to identify CRISPR arrays using <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code>.</p>
</section>
<section id="running-cctk-blast">
<h5>Running <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code><a class="headerlink" href="#running-cctk-blast" title="Permalink to this headline"></a></h5>
<p>First, make a folder to contain the outputs produced by <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code>. Then we can run it. We need to provide a description of an identifier that is present in all the fasta headers for a given assembly as our assemblies are all in multiple contigs. In our case that identifier is the ERR accession we added above. We will provide it as a regex here, but see the <a class="reference internal" href="blast.html#blast-contig-ids"><span class="std std-ref">Associating arrays from different contigs</span></a> section of the <a class="reference external" href="blast.html">cctk blast</a> page for a description of other options for how you can specify this information.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1"># Make output dir</span>
mkdir Blast_CRISPRs

<span class="c1"># Run cctk blast</span>
cctk blast -d Blastdb/assembly_db -r example_repeats.fna -o Blast_CRISPRs/ -p <span class="s2">&quot;ERR\d+&quot;</span>
Total unique spacers: <span class="m">242</span>
Total unique arrays: <span class="m">22</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> produces the same kind of outputs as <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code>. The command line output is a description of how many unique arrays and spacers were identified, and all the same output files are saved to the specified output directory</p>
</section>
</section>
<section id="exploring-crispr-array-relationships-using-a-network-representation">
<span id="network-tutorial"></span><h4>Exploring CRISPR array relationships using a network representation<a class="headerlink" href="#exploring-crispr-array-relationships-using-a-network-representation" title="Permalink to this headline"></a></h4>
<p>Now that we have predicted CRISPR arrays in the example assemblies, we can begin to explore the relationships between these arrays. We will first visualize array relationships as a network to see how arrays in this dataset are related on a broad scale, and then we will explore more closely the relationships between a small number of arrays. In the following example, we will use <a class="reference external" href="https://cytoscape.org/">Cytoscape</a> to visualize our array relationship network and will work with the data we generated using <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code>. We will refer to arrays within the network representation as “nodes” and the relationship between two arrays as an “edge”.</p>
<p>Both <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> and <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> produced a file called <a class="reference internal" href="minced.html#array-network"><span class="std std-ref">Array_network.txt</span></a> which can be read directly in to Cytoscape by simply clicking and dragging the file onto a Cytoscape window. You should then define the columns as in the following image (These definitions are used for applying styles according to node and edge attributes):</p>
<img alt="../_images/network_import_tutorial.png" src="../_images/network_import_tutorial.png" />
<p>After importing the data, we can use styles to colour edges according to the number of spacers shared or the jaccard similarity between arrays to quickly get a sense of the relationships within each of our clusters. Below is an image showing the network with edges coloured according to the Jaccard similarity between each pair of nodes (darker colours indicate higher array similarity).</p>
<img alt="../_images/eg_network_tutorial.png" src="../_images/eg_network_tutorial.png" />
<p>In this network representation of array relationships, it is clear that there are three distinct clusters of arrays. The left-most cluster containing arrays 2, 4, 7, 12, and 17 all have low similarity with one another, while the other two clusters contain some arrays that have a high similarity to one another. Any arrays that do not share any spacers with any other arrays in the dataset are not shown in this network. We will now look more closely at these three clusters using other CCTK tools.</p>
</section>
<section id="using-crisprdiff-to-visualize-array-relationships">
<span id="diffplot-tutorial"></span><h4>Using CRISPRdiff to visualize array relationships<a class="headerlink" href="#using-crisprdiff-to-visualize-array-relationships" title="Permalink to this headline"></a></h4>
<p><a class="reference external" href="crisprdiff.html">CRISPRdiff</a> can be used to quickly and easily identify the spacers that are shared and distinct between CRISPR arrays. Here we will use it to visualise the three clusters of arrays that we saw in the <a class="reference internal" href="#network-tutorial"><span class="std std-ref">Exploring CRISPR array relationships using a network representation</span></a>. For this example we will create a directory within our Minced_CRISPRs/ directory and save plots at that location.</p>
<p><strong>N.B.</strong> In the following sections, the spacers within arrays will be referred to using their index within the array and their colour. e.g. the leader-most (i.e. left-most) spacer in a given array is spacer 1, while the next spacer (2nd spacer) is spacer 2. From the trailer end, spacers will be numbered using negative numbers. E.g. the trailer-most (i.e. right-most) spacer is spacer -1, while the next spacer from the trailer end is -2 etc.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Minced_CRISPRs$ mkdir PLOTS
Minced_CRISPRs$ <span class="nb">cd</span> PLOTS/
</pre></div>
</div>
<p>First let’s look at the left-most cluster in which all arrays have a low level of similarity to one another.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Minced_CRISPRs/PLOTS$ cctk CRISPRdiff -a ../PROCESSED/Array_IDs.txt -o left_cluster.png <span class="m">2</span> <span class="m">4</span> <span class="m">7</span> <span class="m">12</span> <span class="m">17</span>
</pre></div>
</div>
<p>That produces a plot similar to that shown below (the below plot was generated using the additional option <code class="docutils literal notranslate"><span class="pre">--plot-height</span> <span class="pre">1.5</span></code> to reduce vertical spacing).</p>
<img alt="../_images/diff_tutorial_left.png" src="../_images/diff_tutorial_left.png" />
<p>In the above image, it is clear that most spacers are not shared between arrays as they are depicted as thin, black rectangles. This is consistent with the network representation in which all the edges in this cluster have a light colour. In fact, all of the arrays in this cluster share their trailer-most spacer (spacer -1, green), but other spacers are either unique or only shared by a subset of arrays. This pattern may indicate that these arrays come from a common ancestral array that diverged long ago as only the presumably oldest spacer (trailer-most) is shared.</p>
<p>Other interesting relationships are also clear. For example, Arrays 4 and 7 share a black spacer (-5 in array 7, -4 in array 4), which is surrounded on both sides by spacers that are not shared between these two arrays. Some possible explanations for this pattern of spacer sharing will be discussed below in the <a class="reference internal" href="#tree-tutorial"><span class="std std-ref">Using CRISPRtree to create hypotheses of array histories</span></a> section.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Minced_CRISPRs/PLOTS$ cctk CRISPRdiff -a ../PROCESSED/Array_IDs.txt -o left_cluster.png <span class="m">1</span> <span class="m">9</span> <span class="m">11</span> <span class="m">14</span>
</pre></div>
</div>
<img alt="../_images/diff_tutorial_middle.png" src="../_images/diff_tutorial_middle.png" />
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>Minced_CRISPRs/PLOTS$ cctk CRISPRdiff -a ../PROCESSED/Array_IDs.txt -o left_cluster.png <span class="m">3</span> <span class="m">13</span> <span class="m">16</span> <span class="m">18</span>
</pre></div>
</div>
<img alt="../_images/diff_tutorial_right.png" src="../_images/diff_tutorial_right.png" />
</section>
<section id="using-crisprtree-to-create-hypotheses-of-array-histories">
<span id="tree-tutorial"></span><h4>Using CRISPRtree to create hypotheses of array histories<a class="headerlink" href="#using-crisprtree-to-create-hypotheses-of-array-histories" title="Permalink to this headline"></a></h4>
<img alt="../_images/tree_tutorial_left.png" src="../_images/tree_tutorial_left.png" />
</section>
</section>
</section>
<section id="cctk-minced-vs-cctk-blast-output-comparison">
<span id="minced-blast-comp"></span><h2><code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">minced</span></code> vs <code class="docutils literal notranslate"><span class="pre">cctk</span> <span class="pre">blast</span></code> output comparison<a class="headerlink" href="#cctk-minced-vs-cctk-blast-output-comparison" title="Permalink to this headline"></a></h2>
<p>TO ADD</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="spacerblast.html" class="btn btn-neutral float-left" title="spacerblast" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Alan Collins.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>